### PR TITLE
encoding support for vswitch

### LINKED ISSUE #6: vswitch()'s mapping doesn't consider the Encoding issues
``` r
x <- "fa\xE7ile"
Encoding(x) <- "latin1"
x2 <- enc2utf8(x)
kit::vswitch(
  c(x, x2),
  x,
  1
)
#> [1]  1 NA
```

<sup>Created on 2020-10-23 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

<details>

<summary>Session info</summary>

``` r
devtools::session_info()
#> - Session info ---------------------------------------------------------------
#>  setting  value                         
#>  version  R version 4.0.2 (2020-06-22)  
#>  os       Windows 10 x64                
#>  system   x86_64, mingw32               
#>  ui       RTerm                         
#>  language en                            
#>  collate  Chinese (Simplified)_China.936
#>  ctype    Chinese (Simplified)_China.936
#>  tz       Asia/Taipei                   
#>  date     2020-10-23                    
#> 
#> - Packages -------------------------------------------------------------------
#>  package     * version    date       lib source        
#>  assertthat    0.2.1      2019-03-21 [1] CRAN (R 4.0.0)
#>  backports     1.1.7      2020-05-13 [1] CRAN (R 4.0.0)
#>  callr         3.4.3      2020-03-28 [1] CRAN (R 4.0.0)
#>  cli           2.0.2      2020-02-28 [1] CRAN (R 4.0.0)
#>  crayon        1.3.4      2017-09-16 [1] CRAN (R 4.0.0)
#>  desc          1.2.0      2018-05-01 [1] CRAN (R 4.0.0)
#>  devtools      2.3.0      2020-04-10 [1] CRAN (R 4.0.0)
#>  digest        0.6.25     2020-02-23 [1] CRAN (R 4.0.0)
#>  ellipsis      0.3.1      2020-05-15 [1] CRAN (R 4.0.2)
#>  evaluate      0.14       2019-05-28 [1] CRAN (R 4.0.0)
#>  fansi         0.4.1      2020-01-08 [1] CRAN (R 4.0.0)
#>  fs            1.5.0      2020-07-31 [1] CRAN (R 4.0.2)
#>  glue          1.4.2      2020-08-27 [1] CRAN (R 4.0.2)
#>  highr         0.8        2019-03-20 [1] CRAN (R 4.0.0)
#>  htmltools     0.5.0      2020-06-16 [1] CRAN (R 4.0.0)
#>  kit           0.0.4      2020-07-21 [1] CRAN (R 4.0.3)
#>  knitr         1.29       2020-06-23 [1] CRAN (R 4.0.0)
#>  magrittr      1.5        2014-11-22 [1] CRAN (R 4.0.0)
#>  memoise       1.1.0      2017-04-21 [1] CRAN (R 4.0.0)
#>  pkgbuild      1.0.7      2020-04-25 [1] CRAN (R 4.0.0)
#>  pkgload       1.0.2      2018-10-29 [1] CRAN (R 4.0.0)
#>  prettyunits   1.1.1      2020-01-24 [1] CRAN (R 4.0.0)
#>  processx      3.4.3      2020-07-05 [1] CRAN (R 4.0.2)
#>  ps            1.3.4      2020-08-11 [1] CRAN (R 4.0.2)
#>  R6            2.4.1      2019-11-12 [1] CRAN (R 4.0.0)
#>  remotes       2.2.0      2020-07-21 [1] CRAN (R 4.0.2)
#>  rlang         0.4.7      2020-07-09 [1] CRAN (R 4.0.2)
#>  rmarkdown     2.4        2020-09-30 [1] CRAN (R 4.0.2)
#>  rprojroot     1.3-2      2018-01-03 [1] CRAN (R 4.0.0)
#>  sessioninfo   1.1.1      2018-11-05 [1] CRAN (R 4.0.0)
#>  stringi       1.4.6      2020-02-17 [1] CRAN (R 4.0.0)
#>  stringr       1.4.0      2019-02-10 [1] CRAN (R 4.0.0)
#>  testthat      2.3.2.9000 2020-05-09 [1] local         
#>  usethis       1.6.1      2020-04-29 [1] CRAN (R 4.0.0)
#>  withr         2.2.0      2020-04-20 [1] CRAN (R 4.0.0)
#>  xfun          0.18       2020-09-29 [1] CRAN (R 4.0.2)
#>  yaml          2.2.1      2020-02-01 [1] CRAN (R 4.0.0)
#> 
#> [1] D:/app/R_lib/4.0
#> [2] D:/app/R-4.0.0/library
```

</details>

### PR DESCRIPTION
**Pull request by 2005m**
Close #6 , @shrektan please let me know if you happy with that. Thanks

### CODE REVIEW

File: src/vswitch.c [hunk 1]
```diff
@@ -64,6 +68,15 @@ SEXP vswitchR(SEXP x, SEXP values, SEXP outputs, SEXP na, SEXP nthreads) {
   int nprotect=0;
   SEXP ans = R_NilValue;
   SEXPTYPE type_o = UTYPEOF(outputs);
+  SEXP xans = R_NilValue, vans = R_NilValue;
+  bool utfcon = false;
+  if (pchkenc && type_x == STRSXP) {
+    if (isMixEnc(x) || isMixEnc(values) || getCharCE(STRING_PTR(x)[0]) != getCharCE(STRING_PTR(values)[0])) {
```
Comment (shrektan) [inline src/vswitch.c:77]: Here you assume the vector is at least length one. Any chance that a zero-length vector being passed into the routine then leads to segfaults?

Other parts look good to me. Thanks.

Comment (2005m) [reply src/vswitch.c:77]: For the zero length vector, I tried to make the function crash but did not manage to.

Comment (shrektan) [reply src/vswitch.c:77]: If zero-length vectors can reached to these lines, I suggest to handle it carefully. [An example](https://github.com/Rdatatable/data.table/issues/4060#issuecomment-556011718)  is 

```r
Rcpp::cppFunction("int getlength(SEXP x) { return LENGTH(VECTOR_ELT(x,0)); }")
getlength(list())
```

where garbage values generated on R3.4 but always zero on R3.5 - 3.6.

Comment (2005m) [reply src/vswitch.c:77]: I added a check on the length of x and values that should prevent a zero-length vector to reach these lines.

File: src/vswitch.c [hunk 2]
```diff
@@ -37,8 +40,12 @@ SEXP vswitchR(SEXP x, SEXP values, SEXP outputs, SEXP na, SEXP nthreads) {
   SEXPTYPE type_x = UTYPEOF(x);
   SEXPTYPE type_na = UTYPEOF(na);
   SEXPTYPE type_values = UTYPEOF(values);
+  const bool pchkenc = asLogical(chkenc);
   const bool nonna = !isNull(na);
   const bool nalen = (len_na==1);
+  if (!len_values || !len_x) {
+    error("Argument 'x' and 'values' connot be zero-length vector.");
```
Comment (shrektan) [inline src/vswitch.c]: I think 0-length x should return a 0-length value. This is also the current behavior of vswitch().

```r
> kit::vswitch(character(), "a", 1)
numeric(0)
```

Comment (2005m) [reply src/vswitch.c]: above behaviour should remain now

Comment (shrektan) [reply src/vswitch.c]: Great thanks!

**Pull request merged**